src/ooni/utils.cpp

```diff
@@ -13,75 +18,139 @@ void ip_lookup(Callback<Error, std::string> callback, Settings settings,
     ip_lookup_impl(callback, settings, reactor, logger);
 }
 
-static Error geoip_resolve_country (std::string ip, std::string path_country, json &node) { 
-    GeoIP *gi;
-    GeoIPLookup gl;
-    memset (&gl, 0, sizeof(gl));
-    
-    gi = GeoIP_open(path_country.c_str(), GEOIP_MEMORY_CACHE);
-    if (gi == nullptr) {
+void resolver_lookup(Callback<Error, std::string> callback, Settings settings,
+                     Var<Reactor> reactor, Var<Logger> logger) {
+    resolver_lookup_impl(callback, settings, reactor, logger);
+}
+
+IPLocation::IPLocation(std::string path_country, std::string path_asn) {
+    if (path_asn != "") {
+        gi_asn = GeoIP_open(path_asn.c_str(), GEOIP_MEMORY_CACHE);
+    }
```

Comment:
I'd prefer it if we could only cache the paths in the constructor and then lazily open the databases when requested to do so by access APIs. This would allow us to return the proper `ErrorOr<>` error in case it was not possible to open a database, and would also put open and check-after-open in the same place, which I think makes the code more robust to refactoring and easier to defend.